### PR TITLE
Update non-ST FFI to use Fn

### PR DESCRIPTION
**Pull request by JordanMartinez**
**Description of the change**

Updates non-`ST` FFI to use uncurried functions via `Fn` types

---

**Checklist:**

- [x] Added the change to the changelog's "Unreleased" section with a reference to this PR (e.g. "- Made a change (#0000)")
- [ ] Linked any existing issues or proposals that this pull request should close
- [ ] Updated or added relevant documentation
- [ ] Added a test for the contribution (if applicable)

### CODE REVIEW

File: bower.json [hunk 1]
```diff
@@ -35,4 +36,4 @@
     "purescript-const": "^6.0.0",
     "purescript-minibench": "^4.0.0"
   }
-}
+}
```
Comment (thomashoneyman) [inline bower.json:39]:
```suggestion
}

```

**Comment by thomashoneyman** [comment]
What is the motivation for this? Just to write more idiomatic JS in the FFI? Is this a change we're going to apply across core?

**Comment by JordanMartinez** [comment]
Here's my rationale:
1. it produces stylistically cleaner JS output (no test needed)
1. Those doing `dev` work (i.e. just doing `purs compile` w/ no optimizations via `purs-backend-es`) _should_ get code that runs faster (test needed to confirm)
1. once these functions use `Fn*` functions, `purs-backend-es` with no other changes _should_ optimize more things (test needed to confirm)
1. once these functions use `Fn*` functions, `purs-backend-es` can be further updated to know how to evaluate some of these things during its optimization pass (pretty sure this is true).

**Review by thomashoneyman — APPROVED** [review]

**Comment by natefaubion** [comment]
> once these functions use Fn* functions, purs-backend-es with no other changes should optimize more things (test needed to confirm)
once these functions use Fn* functions, purs-backend-es can be further updated to know how to evaluate some of these things during its optimization pass (pretty sure this is true).

Just as an FYI for changes like this, changing FFI calling conventions will potentially break optimizations. `purs-backend-optimizer` cases the calling convention against unsafeIndexImpl (curried vs uncurried matters), so currently the point release here very likely does not optimize at all with backend-es.

**Comment by natefaubion** [comment]
That is, if it's important for the core team (I'm not saying it is) to preserve optimizations in tools like backend-optimizer, it's important that we actually do follow up and verify the impact of these changes.

**Pull request merged**